File: 0645-set-mismatch/0645-set-mismatch.java
public class Solution {
    public int[] findErrorNums(int[] nums) {
        int[] res = new int[2]; // duplicate, missing
        //For each number we found, set nums[number-1] to its negative value (<0)
        for(int i=0; i<nums.length; i++) {
            int idx = Math.abs(nums[i])-1; //since index starts from 0, and the set starts from 1
            if(nums[idx]> 0) nums[idx] = -nums[idx];
            else res[0] = idx+1; //have already been found
        }
        // At this point, only nums[missingNumber-1] > 0
        for(int i=0; i<nums.length; i++) {
            if(nums[i] <0) nums[i] = -nums[i]; //restore the original values
            else res[1]=i+1;//since index starts from 0, and the set starts from 1
        }
        return res;
    }
}

File: 0645-set-mismatch/0645-set-mismatch.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var findErrorNums = function(nums) {
    let repeated, missing;
    for(let i=0; i<nums.length; i++){
        if(nums[Math.abs(nums[i])-1]<0){
            repeated = Math.abs(nums[i]);
        } else {
            nums[Math.abs(nums[i])-1] *= -1;
        }
    }
    for(let i=0; i<nums.length; i++){
        if(nums[i]>0){
            missing = i+1;
        }
    }
    return [repeated, missing];
};